text/0156-columnar-encoding.md

```diff
@@ -0,0 +1,151 @@
+to optimize the processing and the storage of structured data. This proposal aims to leverage this representation to 
+make the Open Telemetry protocol more efficient end-to-end.
+
+> Definition: a multivariate time series has more than one time-dependent variable. Each variable depends not only on 
+its past values but also has some dependency on other variables. A 3 axis accelerometer reporting 3 metrics simultaneously; 
+a mouse move that simultaneously reports the values of x and y, a meteorological weather station reporting temperature, 
+cloud cover, dew point, humidity and wind speed; an http transaction chararterized by many interrelated metrics sharing 
+the same labels are all common examples of multivariate time-series.
+
+This [benchmark](https://github.com/lquerel/otel-multivariate-time-series/blob/main/README2.md) illustrates in detail
+the potential gain we can obtain for a multivariate time-series scenario.
+
+## Explanation
+
+Fundamentally metrics, logs and traces are all structured events occurring at a certain time and optionally covering a 
+specified time span. Creating an efficient and generic representation for events will benefit the entire Open Telemetry 
+eco-system. 
+
+Currently all the OTEL entities are stored in "rows". A metric entity is a protobuf message containing timestamp, labels,
+metric value and few other attributes. A batch of metrics behave as a table with multiple rows or metric messages.
+
+Another way to represent the same data is to represent them in columns instead of rows, i.e. a column containing all the 
+timestamp, a distinct column per label name, a column for the metric values, and so on. This columnar representation is 
+a proven approach to optimize the creation, size, and processing of data batches. The main benefits of a such approach are:
+* better data compression rate (group of similar data) 
+* faster data processing (better data locality => better use of the CPU cache lines)
+* faster serialization and deserialization (less objects to process)
+* faster batch creation (less memory allocation)
+* better IO efficiency (less data to transmit)
+
+The benefit of this approach increases proportionally with the size of the batches. Using the existing "row-oriented" 
+representation is well suited for small batch scenarios. Therefore, this proposal suggests to **extend** the protocol by 
+adding a new columnar event entity to better support multivariate time-series and large batches of events.
+
+## Internal details
+
+In addition to the 3 existing resource entities (metrics, logs and traces), we introduce the resource event entity (see [this protobuf specification](#event-proto)
+for more details) encoding a batch of events into an Apache Arrow buffer. [Apache Arrow](https://arrow.apache.org/) is 
+a language-independent columnar memory format for flat and hierarchical data, organized for efficient analytic operations 
```

Comment:
I would very much like to see an example mapping to Arrow for spans. It is fairly straightforward to map metrics into columnar format, but spans are hierarchical, and I am not familiar with how they can be represented in a columnar format. The benchmark only seems to implement metrics encoding.

Reply:
Arrow supports list of structures so representing attributes, events and links is technically possible (see below for an example of schema including the declaration of attributes). I already have a partial implementation of span in my repo for both the protobuf implementation and the mix protobuf+arrow implementation but I need to find some time to finalize it in order to present the results. 

```rust
Schema {
        Schema::new(vec![
            Field::new("kind", DataType::Int64, true),
            Field::new(
                "status",
                DataType::Struct(vec![
                    Field::new("message", DataType::Utf8, true),
                    Field::new("code", DataType::Int64, true),
                ]),
                true,
            ),
            Field::new("trace_id", DataType::Utf8, false),
            Field::new("span_id", DataType::Utf8, false),
            Field::new("trace_state", DataType::Utf8, true),
            Field::new("parent_span_id", DataType::Utf8, true),
            Field::new("name", DataType::Utf8, false),
            Field::new("attributes", DataType::List(
                Box::new(Field::new(
                    "attr",
                    DataType::Struct(vec![
                        Field::new("name", DataType::Utf8, false),
                        Field::new("value", DataType::Utf8, false),
                    ]),
                    true,
                ))
            ), true),
        ])
 }
```

Reply:
Added a mapping section for each OTEL entities.